src/LabLog/Services/SchoolService.cs

```diff
@@ -14,34 +15,36 @@ public class SchoolService
         private EventModelContext _db;
         private string _user;
         private SchoolEventHandler _school;
+        private string _controllerString;
 
-        public SchoolService(EventModelContext db, string user)
+        public SchoolService(EventModelContext db, string user, string controllerString)
         {
             _db = db;
             _user = user;
             _school = new SchoolEventHandler(_user, db);
+            _controllerString = controllerString;
         }
 
-        public List<SchoolModel> RebuildReadModel()
+        public async Task<List<SchoolModel>> RebuildReadModelAsync()
         {
-            DeleteReadModelFromDatabase();
-            var schoolCreateEvents = _db.LabEvents.Where(o => (o.EventType == SchoolCreatedEvent.EventTypeString));
+            await DeleteReadModelFromDatabaseAsync();
+            List<LabEvent> schoolCreateEvents = await _db.LabEvents.Where(o => (o.EventType == SchoolCreatedEvent.EventTypeString)).ToListAsync();
 
             foreach (ILabEvent e in schoolCreateEvents)
             {
-                SchoolModel school = GetSchool(e.SchoolId);
-                _db.Add(school);
-                _db.SaveChanges();
+                SchoolModel school = await GetSchoolAsync(e.SchoolId);
+                await _db.AddAsync(school);
+                await _db.SaveChangesAsync();
             }
             List<SchoolModel> schools = _db.Schools.ToList();
             return schools;
         }
 
-        private void DeleteReadModelFromDatabase()
+        private async Task DeleteReadModelFromDatabaseAsync()
         {
-            _db.Database.ExecuteSqlCommand("DELETE FROM ComputerModel;");
-            _db.Database.ExecuteSqlCommand("DELETE FROM RoomModel;");
-            _db.Database.ExecuteSqlCommand("DELETE FROM Schools;");
+            await _db.Database.ExecuteSqlCommandAsync("DELETE FROM ComputerModel;");
```

Comment:
You could probably run these three in parallel, rather than waiting for each one to complete.
Rather than doing await, just capture the return object (`Task<something>`) and put it in a list. After calling all 3 and adding all 3 to the list, you can do `await Task.WhenAll(listofTasks)`

Reply:
there's no return object though, so what would the <something> be?

Reply:
Task